piecrust/src/session.rs

```diff
@@ -777,8 +778,20 @@ impl Session {
             })
             .map_err(Error::normalize)?;
         let ret = instance.read_bytes_from_arg_buffer(ret_len as u32);
-
-        let spent = limit - instance.get_remaining_gas();
+        let ret_b = instance.read_from_arg_buffer_b();
+        let allowance = *ret_b.first().unwrap_or(&0u64);
+        let charge = *ret_b.get(1).unwrap_or(&0u64);
+
+        let mut spent = limit - instance.get_remaining_gas();
+
+        if allowance > 0 {
+            // the call has been paid for by the contract so we can set
+            // the spent amount to zero,
+            spent = 0;
+        } else {
+            // add possible contract's charge
+            spent += charge;
+        }
```

Comment:
Wouldn't it make more sense to modify the `CallReceipt` to include what the contract set in the economic buffer? This instead of modifying the gas spent.

Reply:
I evaluated this suggestion and I came to the conclusion that it is important to set the `spent` variable at this particular place. IMO data like allowance or charge cannot go to CallReceipt because this functionality is not optional. System needs to act as if gas was either not consumed (allowance) or consumed with additional charge, so that from the point of view of the caller the behaviour is completely deterministic. If we leave the allowance/charge in `CallReceipt`, we lose control about enforcement of our policy. We cannot be sure if it will be enforced, yet, since we charged/deposited the contract with actual money already at this point, we must make the enforcement final at this point as well.

Reply:
`spent` contains information about gas spent, and any additional data would contain information the economic buffer of a called contract. We don't lose control at all on enforcing any policy currently existing - not sure what you're referring to with this - and allow for more freedom downstream on how to handle it.

Reply:
I do not see any freedom about how to handle it downstream, as the money is already moved at this point, so all we need to do is to even it out, otherwise we may cause economic imbalance - but let's discuss it

Reply:
done